File: test/intencion.test.mjs
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { crearIntencion } from '../src/lib/intencion.js'

/**
 * «Esta posicion la ha elegido el usuario.»
 *
 * Es lo unico que permite subir una posicion anterior a la que hay en la nube,
 * porque el servidor rechaza cualquier retroceso que no venga marcado. Antes se
 * ponia al saltar y solo se bajaba al abrir otro libro, asi que un unico toque
 * a «+30 s» dejaba todas las subidas de las horas siguientes saltandose la
 * comprobacion. Como los botones de salto son los que mas se usan, la
 * proteccion quedaba apagada casi siempre.
 *
 * Misma spec que Intencion.kt.
 */

test('un libro recien abierto no tiene intencion pendiente', () => {
  const i = crearIntencion()
  assert.equal(i.activa, false)
})

test('saltar la marca', () => {
  const i = crearIntencion()
  i.marcar()
  assert.equal(i.activa, true)
})

test('la intencion se CONSUME con la subida que la lleva a la nube', () => {
  // Esto es lo que faltaba: seguia activa el resto de la sesion.
  const i = crearIntencion()
  i.marcar()
  const sello = i.sello
  i.cumplida(sello)
  assert.equal(i.activa, false, 'ya esta guardada: lo que venga despues es inercia')
})

test('una subida fallida no consume la intencion', () => {
  // Si no, un retroceso se perderia por un corte de red: el reintento iria
  // condicionado y el servidor lo rechazaria por ir hacia atras.
  const i = crearIntencion()
  i.marcar()
  const sello = i.sello
  // No se llama a cumplida() porque la subida no llego.
  assert.equal(i.activa, true)
  i.cumplida(sello)
  assert.equal(i.activa, false, 'y se consume cuando por fin llega')
})

test('saltar otra vez mientras la subida esta en vuelo NO pierde la intencion nueva', () => {
  // La carrera: sale la subida de la primera intencion, el usuario salta otra
  // vez, y al volver la respuesta no se puede bajar la bandera porque la que
  // hay ahora es otra y todavia no ha llegado a ninguna parte.
  const i = crearIntencion()
  i.marcar()
  const selloEnVuelo = i.sello

  i.marcar() // el usuario salta de nuevo
  i.cumplida(selloEnVuelo) // responde la subida vieja

  assert.equal(i.activa, true, 'la intencion nueva sigue viva')
  i.cumplida(i.sello)
  assert.equal(i.activa, false)
})

test('varios saltos seguidos cuentan como una sola intencion pendiente', () => {
  const i = crearIntencion()
  i.marcar()
  i.marcar()
  i.marcar()
  assert.equal(i.activa, true)
  i.cumplida(i.sello)
  assert.equal(i.activa, false, 'la ultima subida las salda todas')
})

test('abrir otro libro empieza sin intencion', () => {
  const i = crearIntencion()
  i.marcar()
  i.reiniciar()
  assert.equal(i.activa, false)
})

test('escenario completo: un salto no desactiva la proteccion el resto de la tarde', () => {
  const i = crearIntencion()

  // Pulsas +30 s una vez.
  i.marcar()
  assert.equal(i.activa, true, 'esa subida puede ir hacia atras: la elegiste tu')
  i.cumplida(i.sello)

  // Y sigues escuchando tres horas. Ninguna de esas subidas deberia saltarse
  // la comprobacion del servidor.
  for (let subida = 0; subida < 360; subida++) {
    assert.equal(i.activa, false, `la subida ${subida} debe ir condicionada`)
    i.cumplida(i.sello)
  }
})

File: android/app/src/test/java/com/lumina/audiolibros/sync/IntencionTest.kt
package com.lumina.audiolibros.sync

import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * «Esta posicion la ha elegido el usuario.»
 *
 * Es lo unico que permite subir una posicion anterior a la que hay en la nube,
 * porque el servidor rechaza cualquier retroceso que no venga marcado. Antes se
 * ponia al saltar y solo se bajaba al abrir otro libro, asi que un unico toque
 * a «+30 s» dejaba todas las subidas de las horas siguientes saltandose la
 * comprobacion. Como los botones de salto son los que mas se usan, la
 * proteccion quedaba apagada casi siempre.
 *
 * Misma spec que src/lib/intencion.js.
 */
class IntencionTest {

    @Test
    fun `un libro recien abierto no tiene intencion pendiente`() {
        assertFalse(Intencion().activa())
    }

    @Test
    fun `saltar la marca`() {
        val i = Intencion()
        i.marcar()
        assertTrue(i.activa())
    }

    @Test
    fun `la intencion se consume con la subida que la lleva a la nube`() {
        val i = Intencion()
        i.marcar()
        i.cumplida(i.sello())
        assertFalse("ya esta guardada: lo que venga despues es inercia", i.activa())
    }

    @Test
    fun `una subida fallida no consume la intencion`() {
        // Si no, un retroceso se perderia por un corte de red: el reintento
        // iria condicionado y el servidor lo rechazaria por ir hacia atras.
        val i = Intencion()
        i.marcar()
        val sello = i.sello()
        assertTrue(i.activa())
        i.cumplida(sello)
        assertFalse("y se consume cuando por fin llega", i.activa())
    }

    @Test
    fun `saltar otra vez mientras la subida esta en vuelo no pierde la intencion nueva`() {
        val i = Intencion()
        i.marcar()
        val selloEnVuelo = i.sello()

        i.marcar()                  // el usuario salta de nuevo
        i.cumplida(selloEnVuelo)    // responde la subida vieja

        assertTrue("la intencion nueva sigue viva", i.activa())
        i.cumplida(i.sello())
        assertFalse(i.activa())
    }

    @Test
    fun `varios saltos seguidos cuentan como una sola intencion pendiente`() {
        val i = Intencion()
        i.marcar(); i.marcar(); i.marcar()
        assertTrue(i.activa())
        i.cumplida(i.sello())
        assertFalse("la ultima subida las salda todas", i.activa())
    }

    @Test
    fun `abrir otro libro empieza sin intencion`() {
        val i = Intencion()
        i.marcar()
        i.reiniciar()
        assertFalse(i.activa())
    }

    @Test
    fun `un salto no desactiva la proteccion el resto de la tarde`() {
        val i = Intencion()
        i.marcar()
        assertTrue("esa subida puede ir hacia atras: la eligio el usuario", i.activa())
        i.cumplida(i.sello())

        // Tres horas escuchando: ninguna subida debe saltarse la comprobacion.
        repeat(360) { subida ->
            assertFalse("la subida $subida debe ir condicionada", i.activa())
            i.cumplida(i.sello())
        }
    }
}
